#### This project is about how to predict fit the candidate is based on their available information.

### Importing Librariers

In [1]:
import os
import random 
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df =pd.read_csv("potential-talents - Aspiring human resources - seeking human resources.csv")
df.head()

id                                          job_title  \
0   1  2019 C.T. Bauer College of Business Graduate (...   
1   2  Native English Teacher at EPIK (English Progra...   
2   3              Aspiring Human Resources Professional   
3   4             People Development Coordinator at Ryan   
4   5    Advisory Board Member at Celal Bayar University   

                              location connection  fit  
0                       Houston, Texas         85  NaN  
1                               Kanada      500+   NaN  
2  Raleigh-Durham, North Carolina Area         44  NaN  
3                        Denton, Texas      500+   NaN  
4                       İzmir, Türkiye      500+   NaN

In [3]:
print(f"The data shape is {df.shape}")


The data shape is (104, 5)


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104 entries, 0 to 103
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          104 non-null    int64  
 1   job_title   104 non-null    object 
 2   location    104 non-null    object 
 3   connection  104 non-null    object 
 4   fit         0 non-null      float64
dtypes: float64(1), int64(1), object(3)
memory usage: 4.2+ KB


In [5]:
# The **Fit** column is targeted column soe we can drop it.
df.drop("fit", axis=1, inplace=True)

In [6]:
df.head()

id                                          job_title  \
0   1  2019 C.T. Bauer College of Business Graduate (...   
1   2  Native English Teacher at EPIK (English Progra...   
2   3              Aspiring Human Resources Professional   
3   4             People Development Coordinator at Ryan   
4   5    Advisory Board Member at Celal Bayar University   

                              location connection  
0                       Houston, Texas         85  
1                               Kanada      500+   
2  Raleigh-Durham, North Carolina Area         44  
3                        Denton, Texas      500+   
4                       İzmir, Türkiye      500+

In [7]:
### Checking the numbers of unique value in each columns
for col in df.columns:
    print(f"There is {df[col].nunique()} unique value is {col} column")

There is 104 unique value is id column
There is 52 unique value is job_title column
There is 41 unique value is location column
There is 33 unique value is connection column


In [8]:
print("The number of duplicated row is:", df.duplicated().sum())


The number of duplicated row is: 0


In [9]:
df_dup = df.drop(["id"], axis=1)
print("Number of duplicate entries:" , df_dup.duplicated().sum())

Number of duplicate entries: 51


In [10]:
df[df_dup.duplicated(keep=False)]

id                                          job_title  \
0    1  2019 C.T. Bauer College of Business Graduate (...   
1    2  Native English Teacher at EPIK (English Progra...   
2    3              Aspiring Human Resources Professional   
3    4             People Development Coordinator at Ryan   
4    5    Advisory Board Member at Celal Bayar University   
..  ..                                                ...   
60  61                               HR Senior Specialist   
61  62  Seeking Human Resources HRIS and Generalist Po...   
62  63                      Student at Chapman University   
63  64  SVP, CHRO, Marketing & Communications, CSR Off...   
64  65  Human Resources Coordinator at InterContinenta...   

                               location connection  
0                        Houston, Texas         85  
1                                Kanada      500+   
2   Raleigh-Durham, North Carolina Area         44  
3                         Denton, Texas      500+   
4                        İzmir, Türkiye      500+   
..                                  ...        ...  
60               San Francisco Bay Area      500+   
61            Greater Philadelphia Area      500+   
62              Lake Forest, California          2  
63                  Houston, Texas Area      500+   
64                     Atlanta, Georgia      500+   

[65 rows x 4 columns]

In [11]:
## Dropping the duplicated rows
df_dup =df_dup.drop_duplicates()
df =pd.concat([df["id"], df_dup], axis=1).dropna(axis=0)
print("Shape fo non-duplicated dataframe:", df.shape)

Shape fo non-duplicated dataframe: (53, 4)


In [12]:
df=df.reset_index(drop=True)
df.shape

(53, 4)

In [13]:
df.job_title.value_counts()

Aspiring Human Resources Professional                                                                                    2
2019 C.T. Bauer College of Business Graduate (Magna Cum Laude) and aspiring Human Resources professional                 1
Lead Official at Western Illinois University                                                                             1
Senior Human Resources Business Partner at Heil Environmental                                                            1
Aspiring Human Resources Professional | An energetic and Team-Focused Leader                                             1
HR Manager at Endemol Shine North America                                                                                1
Human Resources professional for the world leader in GIS software                                                        1
RRP Brand Portfolio Executive at JTI (Japan Tobacco International)                                                       1
Information Syst

In [14]:
from collections import Counter

In [15]:
## Checking most frequent words..
words_counts = Counter()
for i in range(len(df)):
    for word in df.job_title[i].split(" "):
        words_counts[word] += 1

print("Number of total words" , len(words_counts))
words_counts.most_common()

Number of total words 221


[('Human', 34),
 ('Resources', 29),
 ('at', 22),
 ('and', 13),
 ('Aspiring', 11),
 ('|', 10),
 ('Professional', 7),
 ('in', 6),
 ('University', 6),
 ('Seeking', 6),
 ('Business', 5),
 ('Student', 5),
 ('Generalist', 5),
 ('Manager', 5),
 ('of', 4),
 ('Specialist', 4),
 ('&', 4),
 ('Management', 4),
 ('seeking', 4),
 ('', 4),
 ('an', 3),
 ('Director', 3),
 ('for', 3),
 ('College', 2),
 ('aspiring', 2),
 ('professional', 2),
 ('Coordinator', 2),
 ('HR', 2),
 ('Senior', 2),
 ('internship', 2),
 ('Resources,', 2),
 ('Staffing', 2),
 ('North', 2),
 ('a', 2),
 ('Resources.', 2),
 ('Major', 2),
 ('to', 2),
 ('Information', 2),
 ('Systems', 2),
 ('-', 2),
 ('Position', 2),
 ('2019', 1),
 ('C.T.', 1),
 ('Bauer', 1),
 ('Graduate', 1),
 ('(Magna', 1),
 ('Cum', 1),
 ('Laude)', 1),
 ('Native', 1),
 ('English', 1),
 ('Teacher', 1),
 ('EPIK', 1),
 ('(English', 1),
 ('Program', 1),
 ('Korea)', 1),
 ('People', 1),
 ('Development', 1),
 ('Ryan', 1),
 ('Advisory', 1),
 ('Board', 1),
 ('Member', 1),
 ('Ce

In [16]:
df.location.value_counts()

Houston, Texas Area                    4
Raleigh-Durham, North Carolina Area    3
Greater New York City Area             3
Austin, Texas Area                     2
Amerika Birleşik Devletleri            2
Kanada                                 2
Greater Philadelphia Area              2
Greater Atlanta Area                   2
Torrance, California                   1
Highland, California                   1
Gaithersburg, Maryland                 1
Baltimore, Maryland                    1
Milpitas, California                   1
Greater Chicago Area                   1
Houston, Texas                         1
Long Beach, California                 1
Chattanooga, Tennessee Area            1
Bridgewater, Massachusetts             1
Lafayette, Indiana                     1
Kokomo, Indiana Area                   1
Las Vegas, Nevada Area                 1
Cape Girardeau, Missouri               1
Greater Los Angeles Area               1
Los Angeles, California                1
Dallas/Fort Wort

### Cleaning the data 
 
    * Removing special character 
    * Writting the text in lower case
    * Removing stopwords and applying lemmatization
    

**Lemmatization usually refers to doing things properly with the use of a vocabulary and morphological analysis of words, normally aiming to remove inflectional endings only and to return the base or dictionary form of a word, which is known as the lemma .**

In [17]:
## Removing special character 
df = df.replace({'location' : { "[\'!#)$%&(*+-./:;<=>?@[\]^_`{|}~\n]" : " "}}, regex=True)
df = df.replace({'connection' : { "[\'!#)$%&(*+-./:;<=>?@[\]^_`{|}~\n]" : " "}}, regex=True)

In [18]:
df.head()

id                                          job_title  \
0   1  2019 C.T. Bauer College of Business Graduate (...   
1   2  Native English Teacher at EPIK (English Progra...   
2   3              Aspiring Human Resources Professional   
3   4             People Development Coordinator at Ryan   
4   5    Advisory Board Member at Celal Bayar University   

                              location connection  
0                       Houston  Texas         85  
1                               Kanada      500    
2  Raleigh Durham  North Carolina Area         44  
3                        Denton  Texas      500    
4                       İzmir  Türkiye      500

In [19]:
df["job_title"] =df["job_title"].str.lower()
df["location"]=df["location"].str.lower()
df.head()

id                                          job_title  \
0   1  2019 c.t. bauer college of business graduate (...   
1   2  native english teacher at epik (english progra...   
2   3              aspiring human resources professional   
3   4             people development coordinator at ryan   
4   5    advisory board member at celal bayar university   

                              location connection  
0                       houston  texas         85  
1                               kanada      500    
2  raleigh durham  north carolina area         44  
3                        denton  texas      500    
4                      i̇zmir  türkiye      500

In [20]:
import nltk
nltk.download("punkt")
nltk.download("wordnet")
nltk.download("stopwords")
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.tokenize.treebank import TreebankWordDetokenizer

import warnings
warnings.filterwarnings('ignore')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\acer\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\acer\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\acer\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [21]:
## Removing stopwords and lemmatization 

def cleaning(df, col):
    stop_words =set(stopwords.words("english"))
    for i in range(len(df)):
        word_token =word_tokenize(df[col][i])
        tokens_without_sw = [w for w in word_token if w not in stop_words]
        lemmatized_sentence = []
        for word in tokens_without_sw:
            lemmatized_sentence.append(WordNetLemmatizer().lemmatize(word))
        df[col][i] =TreebankWordDetokenizer().detokenize(lemmatized_sentence)

In [22]:
print("Job title before removing stopwords:\n", df.job_title.head(1))
print("-"*100)

# Removing stop words
cleaning(df, "job_title")

print("Job title after removing stopwords:\n", df.job_title.head(1))
print("-"*100)
df.head()

Job title before removing stopwords:
 0    2019 c.t. bauer college of business graduate (...
Name: job_title, dtype: object
----------------------------------------------------------------------------------------------------
Job title after removing stopwords:
 0    2019 c.t . bauer college business graduate (ma...
Name: job_title, dtype: object
----------------------------------------------------------------------------------------------------


id                                          job_title  \
0   1  2019 c.t . bauer college business graduate (ma...   
1   2  native english teacher epik (english program k...   
2   3               aspiring human resource professional   
3   4                people development coordinator ryan   
4   5       advisory board member celal bayar university   

                              location connection  
0                       houston  texas         85  
1                               kanada      500    
2  raleigh durham  north carolina area         44  
3                        denton  texas      500    
4                      i̇zmir  türkiye      500

In [23]:
## Replace the abbreviation with their description
df = df.replace({'job_title' : { 'chro' : 'chief human resources officer', 'svp' : 'senior vice president'
        ,'gphr' : 'global professional in human resources','hris' : 'human resources management system'
        , 'csr' : 'corporate social responsibility', 'sphr' : 'strategic and policy-making certification'
        , 'hr' : 'human resources', "[\'!#)$%&(*+-./:;<=>?@[\]^_`{|}~\n]" : "", r'[0-9]' : '', 'epik' : 'tech'
        , 'styczynski lab' : 'tech', 'gi' : 'tech', "schwan's" : 'tech', 'ryan' : 'not tech'
        , 'engine' : 'not tech', 'buckhead atlanta' : 'not tech', 'loparex' : 'not tech', 'delphi Hardware' : 'not tech'
        , "jti" : 'not tech', 'traveler' : 'not tech', 'luxottica' : 'not tech'
        , 'beneteau' : 'not tech', 'scottMadden' : 'not tech', 'ey' : 'not tech', 'endemol' : 'not tech'
        , 'nortia staffing' : 'not tech', 'heil environmental' : 'not tech', 'excellence logging' : 'not tech'}}
        , regex=True)

#### Word Embedding Techniques

I will be using 4 different strategies to find the similarties betwen the targeted sentences and each job title as follows

 1. **TF-IDF**(term frequency-inverse document frequency) is a statistical measure that evaluates how relevant a word is to a document in a collection of documents.
 
    
  2. **GloVe** stands for global vectors for word representation. It is used for generating word embeddings by aggregating global word-word co-occurrence matrix from a corpus. The resulting embeddings show interesting linear substructures of the word in vector space.
    
   
   3. **Word2Vec** is a technique which produces word embeddings for better word representation. We can also say it consists of models for generating word embedding which are shallow two layer neural networks having one input layer, one hidden layer and one output layer.
    
   
   4. **BERT** It is designed to pre-train deep bidirectional representations from unlabeled text by jointly conditioning on both left and right context.

### 1. TF-IDF

In [24]:
import tensorflow as tf
from sklearn.feature_extraction.text import TfidfVectorizer
from scipy.spatial.distance import cosine

In [25]:
## Convert hob titles column into a list
job_title_list =list(df["job_title"])

# Vectorize job_title_list
vectorizer =TfidfVectorizer().fit(job_title_list)
X =vectorizer.transform(job_title_list)

# Get feature names in all the documents
feature_names =vectorizer.get_feature_names()
print("Number of unique features: ", len(feature_names))
print("First 5 features: ", feature_names[:5])

# convert job titles into array
tfidf_vector =X.toarray()
print("Shape of Tfidf Vector: ", tfidf_vector.shape)

Number of unique features:  167
First 5 features:  ['administration', 'administrative', 'admission', 'advisory', 'always']
Shape of Tfidf Vector:  (53, 167)


In [26]:
## Calculating the similarity
def similarity(phrase, column_name):
    # convert search phrase into a vectors
    X = vectorizer.transform(phrase)
    X_vectors = X.toarray()
    print("Shape of search phrase vector:", X_vectors.shape)
    
    # calculate Tfidf cosine similarity 
    sim_score_list = []
    for x in range(len(df)):
        sim_score_list.append(1 - cosine(tfidf_vector[x], X_vectors))
        
    df[column_name] =sim_score_list

In [27]:
# Searched prhase
phrases =pd.DataFrame({"phrase":["aspiring human resources"]})
cleaning(phrases, "phrase")

In [28]:
similarity([phrases.phrase[0]], "TF-IDF_fit_score")

Shape of search phrase vector: (1, 167)


**Let's see the top and lowest matching title job rows with the phrase (aspiring human resources).**

In [29]:
df.sort_values(by ="TF-IDF_fit_score", ascending=False).head()

id                                          job_title  \
2    3               aspiring human resource professional   
45  97               aspiring human resource professional   
5    6                 aspiring human resource specialist   
21  73  aspiring human resource manager seeking intern...   
12  27  aspiring human resource management student see...   

                               location connection  TF-IDF_fit_score  
2   raleigh durham  north carolina area         44          0.781184  
45                 kokomo  indiana area         71          0.781184  
5            greater new york city area          1          0.679890  
21                  houston  texas area          7          0.618600  
12                  houston  texas area      500            0.442293

In [30]:
df.sort_values(by ="TF-IDF_fit_score", ascending=False).tail()

id                                          job_title  \
34   86  information system specialist programmer love ...   
33   85  rrp brand portfolio executive not tech japan t...   
28   80            junior me entechneer information system   
1     2  native english teacher tech english program korea   
52  104       director administration excellence logtechng   

                             location connection  TF-IDF_fit_score  
34             gaithersburg  maryland          4               0.0  
33          greater philadelphia area      500                 0.0  
28  myrtle beach  south carolina area         52               0.0  
1                              kanada      500                 0.0  
52                        katy  texas      500                 0.0

### 2. GloVe

In [31]:
#!pip install wget
#import wget
#wget.download('http://nlp.stanford.edu/data/glove.6B.zip')

In [32]:
#import zipfile as zf
#files = zf.ZipFile("glove.6B.zip", 'r')
#files.extractall('GloVe')
#files.close()

In [33]:
import gensim.downloader as api
from gensim.test.utils import get_tmpfile
from gensim.scripts.glove2word2vec import glove2word2vec
from gensim.models import KeyedVectors, Word2Vec

# Create temp file and save converted embedding into it
target_file = get_tmpfile("word2vec.6B.50d.txt")
glove2word2vec("GloVe/glove.6B.50d.txt", target_file)

# Load the converetd embedding into memory
glove_model = KeyedVectors.load_word2vec_format(target_file)

# Save as binary data
glove_model.save_word2vec_format('word2vec.6B.50d.bin.gz', binary=True)

In [34]:
def doc_token_vectors(sentence, model, sentence_vector_list, vector_dimensions):
    word_tokens = word_tokenize(sentence)
    filtered_words = [w for w in word_tokens if w in model]
    for j in range(len(word_tokens)):
        if word_tokens[j] in filtered_words:
            token_vector = model[word_tokens[j]]
        else:
            token_vector = np.zeros(vector_dimensions)
        sentence_vector_list.append(token_vector)
    
    return sentence_vector_list

In [35]:
def model_fitt_score(df, col, model, vector_dimensions, phrase, fitt_col_name):
    # Vectorize job title using the model
    model_vectors = []
    for i in range(len(df)):
        model_sentence_vector = []
        doc_token_vectors(df[col][i], model, model_sentence_vector, vector_dimensions)
        model_vectors.append(model_sentence_vector)

    print('model vectors shape', np.shape(model_vectors))
    # Vectorize searched phrase using the model
    model_search_phrase_vector = []
    doc_token_vectors(phrase, model, model_search_phrase_vector, vector_dimensions)
    print('model search phrase vector shape', np.shape(model_search_phrase_vector))

    # Calculate cosine similarity between searched phrase and job title
    model_similarity =[]
    for i in range(len(df)):
        sim_score = 1 - cosine(np.mean(model_vectors[i], axis = 0), np.mean(model_search_phrase_vector, axis =0))
        model_similarity.append(sim_score)

    # Add model similarity score to the pt dataframe
    df[fitt_col_name] = model_similarity

In [36]:
model_fitt_score(df, 'job_title', glove_model, 50, phrases.phrase[0], 'GloVe_fit_score')

model vectors shape (53,)
model search phrase vector shape (3, 50)


In [37]:
df.sort_values(by="GloVe_fit_score", ascending =False).head()

id                                          job_title  \
5    6                 aspiring human resource specialist   
2    3               aspiring human resource professional   
45  97               aspiring human resource professional   
21  73  aspiring human resource manager seeking intern...   
22  74                        human resource professional   

                               location connection  TF-IDF_fit_score  \
5            greater new york city area          1          0.679890   
2   raleigh durham  north carolina area         44          0.781184   
45                 kokomo  indiana area         71          0.781184   
21                  houston  texas area          7          0.618600   
22                  greater boston area         16          0.421758   

    GloVe_fit_score  
5          0.969614  
2          0.958512  
45         0.958512  
21         0.934672  
22         0.933199

### 3.Word2Vec

In [38]:
GoogleNews_model = api.load('word2vec-google-news-300')

In [39]:
model_fitt_score(df, 'job_title', GoogleNews_model, 300, phrases.phrase[0], 'GoogleNews_fit_score')

model vectors shape (53,)
model search phrase vector shape (3, 300)


In [40]:
df.sort_values(by ='GoogleNews_fit_score', ascending = False).head()

id                                          job_title  \
45  97               aspiring human resource professional   
2    3               aspiring human resource professional   
5    6                 aspiring human resource specialist   
21  73  aspiring human resource manager seeking intern...   
22  74                        human resource professional   

                               location connection  TF-IDF_fit_score  \
45                 kokomo  indiana area         71          0.781184   
2   raleigh durham  north carolina area         44          0.781184   
5            greater new york city area          1          0.679890   
21                  houston  texas area          7          0.618600   
22                  greater boston area         16          0.421758   

    GloVe_fit_score  GoogleNews_fit_score  
45         0.958512              0.950395  
2          0.958512              0.950395  
5          0.969614              0.912262  
21         0.934672              0.875945  
22         0.933199              0.874494

## 4. BERT

In [46]:
#!pip install sentence_transformers

In [50]:
from sentence_transformers import SentenceTransformer
# build bert_base model
bert_model = SentenceTransformer("bert-base-nli-mean-tokens")

Downloading:   0%|          | 0.00/391 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/190 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/3.95k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/625 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/122 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/229 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/438M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/466k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/399 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/232k [00:00<?, ?B/s]

In [51]:
# Convert job title into BERT embedded vectors
bert_job_title_embeddings =bert_model.encode(list(df.job_title))
bert_job_title_embeddings.shape

(53, 768)

In [52]:
# convert search phrase into a bert embedded vector
bert_search_phrase_embeddings = bert_model.encode(phrases.phrase[0])
bert_search_phrase_embeddings.shape

(768,)

In [53]:
# Calculate cosine similarity between job title and search phrase vectors
bert_cosine_similarity= []
for i in range(len(df)):
    cos_sim = 1-cosine(bert_job_title_embeddings[i], bert_search_phrase_embeddings)
    bert_cosine_similarity.append(cos_sim)
    

# add bert cosine_simirality column in th dataframe
df["BERT_model_fit_score"] =bert_cosine_similarity


In [54]:
df.sort_values(by="BERT_model_fit_score", ascending=False).head()

id                                          job_title  \
5    6                 aspiring human resource specialist   
2    3               aspiring human resource professional   
45  97               aspiring human resource professional   
30  82  aspiring human resource professional  energeti...   
47  99                    seeking human resource position   

                               location connection  TF-IDF_fit_score  \
5            greater new york city area          1          0.679890   
2   raleigh durham  north carolina area         44          0.781184   
45                 kokomo  indiana area         71          0.781184   
30                   austin  texas area        174          0.379605   
47               las vegas  nevada area         48          0.289466   

    GloVe_fit_score  GoogleNews_fit_score  BERT_model_fit_score  
5          0.969614              0.912262              0.955137  
2          0.958512              0.950395              0.948828  
45         0.958512              0.950395              0.948828  
30         0.895589              0.827266              0.867910  
47         0.869983              0.728513              0.849293

In [55]:
df.sort_values(by="BERT_model_fit_score", ascending=False).tail()

id                                          job_title  \
44   96  student indiana university kokomo  business ma...   
50  102           business intelligence analytics not tech   
41   93  admission representative community medical cen...   
33   85  rrp brand portfolio executive not tech japan t...   
35   87  bachelor science biology victoria university w...   

                      location connection  TF-IDF_fit_score  GloVe_fit_score  \
44          lafayette  indiana         19               0.0         0.517594   
50  greater new york city area         49               0.0         0.662951   
41      long beach  california          9               0.0         0.664062   
33   greater philadelphia area      500                 0.0         0.579212   
35         baltimore  maryland         40               0.0         0.518551   

    GoogleNews_fit_score  BERT_model_fit_score  
44              0.314945              0.230208  
50              0.344934              0.198931  
41              0.262623              0.181031  
33              0.271064              0.147233  
35              0.277453              0.136410

**Among all of the models, the BERT model gives us a high score for the top rows and a low one for unrelated titles.**

## Learning to Rank (LTR)

In [56]:
df.head()

id                                          job_title  \
0   1   ct  bauer college business graduate magna cum...   
1   2  native english teacher tech english program korea   
2   3               aspiring human resource professional   
3   4            people development coordinator not tech   
4   5       advisory board member celal bayar university   

                              location connection  TF-IDF_fit_score  \
0                       houston  texas         85          0.254004   
1                               kanada      500            0.000000   
2  raleigh durham  north carolina area         44          0.781184   
3                        denton  texas      500            0.000000   
4                      i̇zmir  türkiye      500            0.000000   

   GloVe_fit_score  GoogleNews_fit_score  BERT_model_fit_score  
0         0.627218              0.592438              0.557126  
1         0.650306              0.283308              0.408490  
2         0.958512              0.950395              0.948828  
3         0.734886              0.358650              0.318839  
4         0.450873              0.206563              0.461810

In [57]:
star_candidate=input("Do you want to star any candidates? Enter 'Yes' or 'No': ")

starred= []
if star_candidate.lower()=="yes":
    starred=[int(item) for item in input("Enter ids of candidates you want to star (separated by space) : ").split()]

Do you want to star any candidates? Enter 'Yes' or 'No': yes
Enter ids of candidates you want to star (separated by space) : 8


In [58]:
df["starred_score"] =df["BERT_model_fit_score"]
for id_num in starred:
    df.loc[df['id'] == id_num, "starred_score"]=1
df.head()

id                                          job_title  \
0   1   ct  bauer college business graduate magna cum...   
1   2  native english teacher tech english program korea   
2   3               aspiring human resource professional   
3   4            people development coordinator not tech   
4   5       advisory board member celal bayar university   

                              location connection  TF-IDF_fit_score  \
0                       houston  texas         85          0.254004   
1                               kanada      500            0.000000   
2  raleigh durham  north carolina area         44          0.781184   
3                        denton  texas      500            0.000000   
4                      i̇zmir  türkiye      500            0.000000   

   GloVe_fit_score  GoogleNews_fit_score  BERT_model_fit_score  starred_score  
0         0.627218              0.592438              0.557126       0.557126  
1         0.650306              0.283308              0.408490       0.408490  
2         0.958512              0.950395              0.948828       0.948828  
3         0.734886              0.358650              0.318839       0.318839  
4         0.450873              0.206563              0.461810       0.461810

In [65]:
import torch
import torch.nn as nn


# Set a seed value
seed_value= 12321 

# 1. Set `PYTHONHASHSEED` environment variable at a fixed value
os.environ['PYTHONHASHSEED']=str(seed_value)

# 2. Set `python` built-in pseudo-random generator at a fixed value
random.seed(seed_value)

# 3. Set `numpy` pseudo-random generator at a fixed value
np.random.seed(seed_value)

# 4. Set `torch` pseudo-random generator at a fixed value
torch.manual_seed(seed_value)

In [66]:
class RankNet(nn.Module):
    
    def __init__(self, num_feature):
        super(RankNet, self).__init__()

        self.model = nn.Sequential(
            nn.Linear(num_feature, 512),         
            nn.Dropout(0.5),                     
            nn.LeakyReLU(0.2, inplace=True),     
            nn.Linear(512, 256),
            nn.Dropout(0.5),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Linear(256, 1),
            nn.Sigmoid())
        
        self.output_sig = nn.Sigmoid()

    def forward(self, input_1, input_2):
        s1 = self.model(input_1)
        s2 = self.model(input_2)
        out = self.output_sig(s1-s2)
        return out
    
    def predict(self, input_):
        s = self.model(input_)
        return s

In [68]:
random_row_1 = df.sample(n=100, replace=True)
random_row_2 =df.sample(n=100, replace=True)
job_title_list_ranknet1 = list(random_row_1["job_title"])
job_title_list_ranknet2 = list(random_row_2["job_title"])
doc_1 =bert_model.encode(job_title_list_ranknet1)
doc_2 =bert_model.encode(job_title_list_ranknet2)
doc_1 =torch.from_numpy(doc_1).float()
doc_2 = torch.from_numpy(doc_2).float()

In [69]:
y_1 = list(random_row_1['starred_score'])
y_2 = list(random_row_2['starred_score'])
y = torch.tensor([1.0 if y1_i>y2_i else 0.5 if y1_i==y2_i else 0.0 for y1_i, y2_i in zip(y_1, y_2)]).float()

y = y.unsqueeze(1)

In [71]:
def train_model(optim, lr_list, epoch, patience):
    # track the training loss as the model trains
    losses= []
    
    model =RankNet(num_feature=768)
    loss_fun= torch.nn.BCELoss()
    
    for lr in lr_list:
        if optim =="SGD":
            optimizer =torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
        elif optim=="Adam":
            optimizer =torch.optim.Adam(model.parameters(), lr=lr)
        print("lr: ", lr, "optimizer: ", optim)
        epoch = epoch
        train_losses  =[]
        valid_losses = []
        
        for i in range(1, epoch+1):
            model.zero_grad()
            y_pred =model(doc_1, doc_2)
            loss=loss_fun(y_pred, y)
            loss.backward()
            optimizer.step()
            losses.append(loss.item())
            
            if i %100 ==0:
                print("Epoch{}, loss: {}".format(i, loss.item()))
                
        print("="*100)
    return model, losses

In [72]:
model, train_loss= train_model(optim="SGD", 
                              lr_list=[0.1,0.01,0.001,0.0001,0.00001],
                              epoch=1000,
                              patience=20)

lr:  0.1 optimizer:  SGD
Epoch100, loss: 0.5080745220184326
Epoch200, loss: 0.4885023832321167
Epoch300, loss: 0.4944716691970825
Epoch400, loss: 0.4932858943939209
Epoch500, loss: 0.4906902313232422
Epoch600, loss: 0.49168312549591064
Epoch700, loss: 0.49649161100387573
Epoch800, loss: 0.48243194818496704
Epoch900, loss: 0.48973309993743896
Epoch1000, loss: 0.4859377145767212
lr:  0.01 optimizer:  SGD
Epoch100, loss: 0.4868333339691162
Epoch200, loss: 0.4887847602367401
Epoch300, loss: 0.48420047760009766
Epoch400, loss: 0.4925222098827362
Epoch500, loss: 0.48924899101257324
Epoch600, loss: 0.48541712760925293
Epoch700, loss: 0.48472148180007935
Epoch800, loss: 0.487932026386261
Epoch900, loss: 0.4922143816947937
Epoch1000, loss: 0.48818713426589966
lr:  0.001 optimizer:  SGD
Epoch100, loss: 0.4889654815196991
Epoch200, loss: 0.48435497283935547
Epoch300, loss: 0.4921264350414276
Epoch400, loss: 0.48823028802871704
Epoch500, loss: 0.48427486419677734
Epoch600, loss: 0.4934704899787903

In [73]:
model, train_loss =train_model(optim="SGD", lr_list=[0.01],
                              epoch=2000,
                              patience=20)

lr:  0.01 optimizer:  SGD
Epoch100, loss: 0.5159167647361755
Epoch200, loss: 0.5027689933776855
Epoch300, loss: 0.4991055727005005
Epoch400, loss: 0.4962245523929596
Epoch500, loss: 0.4974112808704376
Epoch600, loss: 0.4880216717720032
Epoch700, loss: 0.4875315725803375
Epoch800, loss: 0.4951327443122864
Epoch900, loss: 0.488869845867157
Epoch1000, loss: 0.49069511890411377
Epoch1100, loss: 0.4848949909210205
Epoch1200, loss: 0.4906507134437561
Epoch1300, loss: 0.4879225194454193
Epoch1400, loss: 0.49273258447647095
Epoch1500, loss: 0.4900534749031067
Epoch1600, loss: 0.4932117462158203
Epoch1700, loss: 0.49330803751945496
Epoch1800, loss: 0.48913443088531494
Epoch1900, loss: 0.48739299178123474
Epoch2000, loss: 0.4885859787464142


In [74]:
model, train_loss =train_model(optim="Adam",
                              lr_list=[0.1,0.01,0.001,0.0001,0.00001],
                              epoch=1000,
                              patience=20)

lr:  0.1 optimizer:  Adam
Epoch100, loss: 0.7025609016418457
Epoch200, loss: 0.6931471228599548
Epoch300, loss: 0.6893482804298401
Epoch400, loss: 0.6931471228599548
Epoch500, loss: 0.6931471228599548
Epoch600, loss: 0.6893482804298401
Epoch700, loss: 0.6931471228599548
Epoch800, loss: 0.6931471228599548
Epoch900, loss: 0.6931471228599548
Epoch1000, loss: 0.6931471228599548
lr:  0.01 optimizer:  Adam
Epoch100, loss: 0.6931471228599548
Epoch200, loss: 0.6931471228599548
Epoch300, loss: 0.6931471228599548
Epoch400, loss: 0.6893482804298401
Epoch500, loss: 0.6893482804298401
Epoch600, loss: 0.6893482804298401
Epoch700, loss: 0.6931471228599548
Epoch800, loss: 0.6931471228599548
Epoch900, loss: 0.6931471228599548
Epoch1000, loss: 0.6931471228599548
lr:  0.001 optimizer:  Adam
Epoch100, loss: 0.6931471228599548
Epoch200, loss: 0.6931471228599548
Epoch300, loss: 0.6893482208251953
Epoch400, loss: 0.6917506456375122
Epoch500, loss: 0.6855493783950806
Epoch600, loss: 0.6893482208251953
Epoch70

In [75]:
class RankNet(nn.Module):
    
    def __init__(self, num_feature):
        super(RankNet, self).__init__()
        
        self.model = nn.Sequential(
            nn.Linear(num_feature, num_feature*2),
            nn.Dropout(0.2),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Linear(num_feature*2, num_feature *4),
            nn.Dropout(0.2),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Linear(num_feature*4,1),
            nn.Sigmoid())
        self.output_sig =nn.Sigmoid()
        
    def forward(self, input_1, input_2):
        s1 =self.model(input_1)
        s2 =self.model(input_2)
        out =self.output_sig(s1-s2)
        return out
    
    def predict(self, input_):
        s = self.model(input_)
        return s

In [76]:
model, train_loss =train_model(optim="SGD", 
                              lr_list=[0.01],
                              epoch=2000,
                              patience=20)

lr:  0.01 optimizer:  SGD
Epoch100, loss: 0.5054026246070862
Epoch200, loss: 0.4938727617263794
Epoch300, loss: 0.49243098497390747
Epoch400, loss: 0.49046391248703003
Epoch500, loss: 0.4883538782596588
Epoch600, loss: 0.4897862374782562
Epoch700, loss: 0.487813264131546
Epoch800, loss: 0.48874035477638245
Epoch900, loss: 0.4877861440181732
Epoch1000, loss: 0.4909791052341461
Epoch1100, loss: 0.4902816116809845
Epoch1200, loss: 0.48854759335517883
Epoch1300, loss: 0.49258536100387573
Epoch1400, loss: 0.48743969202041626
Epoch1500, loss: 0.4873252809047699
Epoch1600, loss: 0.48870840668678284
Epoch1700, loss: 0.4916936755180359
Epoch1800, loss: 0.48628273606300354
Epoch1900, loss: 0.4872210621833801
Epoch2000, loss: 0.48841214179992676


In [77]:
pred_score = []
for i in range(len(df)):
    embedding =bert_model.encode([df["job_title"][i]])
    embedding_tensor = torch.from_numpy(embedding).float()
    pred = round(model.predict(embedding_tensor).detach().numpy().sum(),2)
    pred_score.append(pred)
    
    

In [78]:
df["RankNet_score"] =pred_score
df.sort_values(by="RankNet_score", ascending =False).head()

id                                          job_title  \
52  104       director administration excellence logtechng   
22   74                        human resource professional   
24   76  aspiring human resource professional  passiona...   
27   79  liberal art major  aspiring human resource ana...   
30   82  aspiring human resource professional  energeti...   

                       location connection  TF-IDF_fit_score  GloVe_fit_score  \
52                  katy  texas      500            0.000000         0.665444   
22          greater boston area         16          0.421758         0.933199   
24           new york  new york        212          0.259969         0.917865   
27  baton rouge  louisiana area          7          0.354417         0.850251   
30           austin  texas area        174          0.379605         0.895589   

    GoogleNews_fit_score  BERT_model_fit_score  starred_score  RankNet_score  
52              0.191680              0.621203       0.621203            1.0  
22              0.874494              0.809261       0.809261            1.0  
24              0.741262              0.765864       0.765864            1.0  
27              0.731783              0.557962       0.557962            1.0  
30              0.827266              0.867910       0.867910            1.0

## CONCLUSION



`BERT` Model was one of the best model to find similarity betwwen our data and the targeted phrase. As for the Ranking Model I ran `RankNet` model on the data and the best Loss score was 48%